File: runtime/ios/Sources/GSXNativeKit/GSXBridge.swift
import Foundation

public struct GSXCapabilitySpec: Codable, Equatable {
    public var name: String
    public var targets: [String]
    public var required: Bool

    public init(name: String, targets: [String] = [], required: Bool = false) {
        self.name = name
        self.targets = targets
        self.required = required
    }

    public func supports(target: String?) -> Bool {
        guard let target, !target.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return true
        }
        let normalized = target.lowercased()
        return targets.isEmpty || targets.map { $0.lowercased() }.contains(normalized)
    }
}

public struct GSXCapabilityReport: Equatable {
    public var required: [String]
    public var available: Set<String>
    public var missing: [String]

    public init(required: [String], available: Set<String>, missing: [String]) {
        self.required = required
        self.available = available
        self.missing = missing
    }

    public var isSatisfied: Bool {
        missing.isEmpty
    }
}

public enum GSXCapabilityChecker {
    public static func check(
        required specs: [GSXCapabilitySpec],
        available: Set<String>,
        target: String? = nil
    ) -> GSXCapabilityReport {
        let normalizedAvailable = Set(available.map { $0.lowercased() })
        let requiredNames = specs
            .filter { $0.required && $0.supports(target: target) }
            .map(\.name)
            .sorted()
        let missing = requiredNames
            .filter { !normalizedAvailable.contains($0.lowercased()) }

        return GSXCapabilityReport(required: requiredNames, available: normalizedAvailable, missing: missing)
    }
}

public final class GSXBridgeClient {
    private let dataClient: GSXDataClient

    public init(dataClient: GSXDataClient) {
        self.dataClient = dataClient
    }

    public convenience init(transport: any GSXTransport) {
        self.init(dataClient: GSXDataClient(transport: transport))
    }

    public convenience init(baseURL: URL, defaultHeaders: [String: String] = [:]) {
        self.init(dataClient: GSXDataClient(baseURL: baseURL, defaultHeaders: defaultHeaders))
    }

    public convenience init(baseURL: URL, defaultHeaders: [String: String] = [:], tokenStore: any GSXTokenStore) {
        self.init(dataClient: GSXDataClient(baseURL: baseURL, defaultHeaders: defaultHeaders, tokenStore: tokenStore))
    }

    public convenience init(baseURL: String, defaultHeaders: [String: String] = [:]) throws {
        self.init(dataClient: try GSXDataClient(baseURL: baseURL, defaultHeaders: defaultHeaders))
    }

    public convenience init(baseURL: String, defaultHeaders: [String: String] = [:], tokenStore: any GSXTokenStore) throws {
        self.init(dataClient: try GSXDataClient(baseURL: baseURL, defaultHeaders: defaultHeaders, tokenStore: tokenStore))
    }

    public func call(_ request: GSXRequest, policy: GSXRequestPolicy = GSXRequestPolicy()) async throws -> GSXResponse {
        try await dataClient.submit(request, policy: policy)
    }
}

File: runtime/android/src/main/kotlin/com/gosx/nativekit/GSXBridge.kt
package com.gosx.nativekit

data class GSXCapabilitySpec(
    val name: String,
    val targets: List<String> = emptyList(),
    val required: Boolean = false,
) {
    fun supports(target: String?): Boolean {
        val normalized = target?.trim()?.lowercase().orEmpty()
        if (normalized.isBlank()) {
            return true
        }
        return targets.isEmpty() || targets.map { it.lowercase() }.contains(normalized)
    }
}

data class GSXCapabilityReport(
    val required: List<String>,
    val available: Set<String>,
    val missing: List<String>,
) {
    val isSatisfied: Boolean
        get() = missing.isEmpty()
}

object GSXCapabilityChecker {
    fun check(
        required: List<GSXCapabilitySpec>,
        available: Set<String>,
        target: String? = null,
    ): GSXCapabilityReport {
        val normalizedAvailable = available.map { it.lowercase() }.toSet()
        val requiredNames = required
            .filter { it.required && it.supports(target) }
            .map { it.name }
            .sorted()
        val missing = requiredNames.filter { !normalizedAvailable.contains(it.lowercase()) }

        return GSXCapabilityReport(
            required = requiredNames,
            available = normalizedAvailable,
            missing = missing,
        )
    }
}

class GSXBridgeClient(
    private val dataClient: GSXDataClient,
) {
    constructor(transport: GSXTransport) : this(GSXDataClient(transport))

    constructor(baseURL: String, defaultHeaders: Map<String, String> = emptyMap()) : this(
        GSXDataClient(baseURL = baseURL, defaultHeaders = defaultHeaders),
    )

    constructor(baseURL: String, defaultHeaders: Map<String, String> = emptyMap(), tokenStore: GSXTokenStore) : this(
        GSXDataClient(baseURL = baseURL, defaultHeaders = defaultHeaders, tokenStore = tokenStore),
    )

    suspend fun call(
        request: GSXRequest,
        policy: GSXRequestPolicy = GSXRequestPolicy(),
    ): GSXResponse = dataClient.submit(request, policy = policy)
}
